File: src/leetcode/java/MinimumAbsoluteDifferenceInBST_530.java
package leetcode.java;

public class MinimumAbsoluteDifferenceInBST_530 {
  private TreeNode prev = new TreeNode(Integer.MAX_VALUE);

  public int getMinimumDifference(TreeNode root) {
    return getMin(root, Integer.MAX_VALUE);
  }

  private int getMin(TreeNode node, int preMin) {
    if (node == null) return Math.min(preMin, Integer.MAX_VALUE);
    int min = getMin(node.left, preMin);
    min = Math.min(min, Math.abs(node.val - prev.val));
    prev = node;
    return Math.min(min, getMin(node.right, min));
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}

File: src/leetcode/python/MinimumAbsoluteDifferenceInBST_530.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class MinimumAbsoluteDifferenceInBST_530(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def inorder(node):
            return inorder(node.left) + [node.val] + inorder(node.right) if node else []

        n = inorder(root)
        return min([n[i + 1] - n[i] for i in range(len(n) - 1)])
